src/Raven.Client/Document/Subscription.cs

```diff
@@ -32,289 +41,444 @@ namespace Raven.Client.Document
-        private readonly Func<Task> ensureOpenSubscription;
-        private readonly ConcurrentSet<IObserver<T>> subscribers = new ConcurrentSet<IObserver<T>>();
-        private readonly SubscriptionConnectionOptions options;
         private readonly CancellationTokenSource cts = new CancellationTokenSource();
+        private readonly Func<Task> ensureOpenSubscription;
         private readonly GenerateEntityIdOnTheClient generateEntityIdOnTheClient;
-        private readonly bool isStronglyTyped;
         private readonly long id;
-        private Task pullingTask;
-        private Task startPullingTask;
-        private IDisposable putDocumentsObserver;
-        private IDisposable endedBulkInsertsObserver;
-        private IDisposable dataSubscriptionReleasedObserver;
+        private readonly bool isStronglyTyped;
+        private readonly SubscriptionConnectionOptions options;
+        private readonly ConcurrentSet<IObserver<T>> subscribers = new ConcurrentSet<IObserver<T>>();
+        private readonly RavenClientWebSocket webSocket;
         private bool completed;
+        private IDisposable dataSubscriptionReleasedObserver;
         private bool disposed;
+        private IDisposable endedBulkInsertsObserver;
         private bool firstConnection = true;
+        private Task pullingTask;
+        private Task startPullingTask;
 
-        public event BeforeBatch BeforeBatch = delegate { };
-        public event AfterBatch AfterBatch = delegate { };
-        public event BeforeAcknowledgment BeforeAcknowledgment = () => true;
-        public event AfterAcknowledgment AfterAcknowledgment = delegate { };
-
-        internal Subscription(long id, string database, SubscriptionConnectionOptions options, IAsyncDatabaseCommands commands, IDatabaseChanges changes, DocumentConvention conventions, bool open, Func<Task> ensureOpenSubscription)
+        internal Subscription(long id, string database, SubscriptionConnectionOptions options,
+            IAsyncDatabaseCommands commands, DocumentConvention conventions, bool open,
+            Func<Task> ensureOpenSubscription)
         {
             this.id = id;
             this.options = options;
             this.commands = commands;
-            this.changes = changes;
             this.conventions = conventions;
             this.ensureOpenSubscription = ensureOpenSubscription;
+            webSocket = new RavenClientWebSocket();
 
             if (typeof(T) != typeof(RavenJObject))
             {
                 isStronglyTyped = true;
-                generateEntityIdOnTheClient = new GenerateEntityIdOnTheClient(conventions, entity => AsyncHelpers.RunSync(() => conventions.GenerateDocumentKeyAsync(database, commands, entity)));
+                generateEntityIdOnTheClient = new GenerateEntityIdOnTheClient(conventions,
+                    entity =>
+                        AsyncHelpers.RunSync(() => conventions.GenerateDocumentKeyAsync(database, commands, entity)));
```

Comment:
We are better off throwing here. We are never actually going to generate an id here.